### PR TITLE
RWA-177: Search Sorting

### PR DESCRIPTION
**Pull request by alexbaeza**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RWA-177

### Change description ###

Added sorting filters

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetTaskByIdControllerTest.java [hunk 1]
```diff
@@ -88,7 +88,7 @@ public void should_return_a_200_and_retrieve_a_task_by_id_jurisdiction_location_
             .and().contentType(MediaType.APPLICATION_JSON_VALUE)
             .and().body("task.id", equalTo(taskId));
 
-        common.cleanUpTask(taskId, REASON_COMPLETED);
+        //common.cleanUpTask(taskId, REASON_COMPLETED);
```
Comment (iguannago) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetTaskByIdControllerTest.java:91]: if not needed, could we clean that up?

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostTaskSearchControllerTest.java [hunk 2]
```diff
@@ -251,6 +258,156 @@ public void should_return_a_200_with_empty_search_results_user_jurisdiction_perm
...
+        result.then().assertThat()
+            .statusCode(HttpStatus.OK.value())
+            .body("tasks.id", hasItems(tasksCreated.get(0).get("taskId"), tasksCreated.get(1).get("taskId")))
+            .body("tasks.case_id", hasItems(tasksCreated.get(0).get("caseId"), tasksCreated.get(1).get("caseId")))
+            .body("tasks.task_state", everyItem(either(is("unassigned")).or(is("assigned"))))
+            .body("tasks.location", everyItem(equalTo("765324")))
+            .body("tasks.jurisdiction", everyItem(equalTo("IA")));
+
+        List<String> jsonResponse = result.jsonPath().getList("tasks.case_id");
+        assertTrue(Ordering.natural().reverse().isOrdered(jsonResponse));
```
Comment (iguannago) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostTaskSearchControllerTest.java:291]: This ordering approach looks good to me. We might want to write a simple unit test to cover this line is isolation and confirm it works as we expect.

**Review by iguannago — APPROVED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetTaskByIdControllerTest.java (hunk 1 above)
Comment (alexbaeza) [reply src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetTaskByIdControllerTest.java:91]: Oops looks like a typo will revert

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostTaskSearchControllerTest.java (hunk 2 above)
Comment (alexbaeza) [reply src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostTaskSearchControllerTest.java:291]: good idea I'll add an extra unit test for confidence

**Review by iguannago — DISMISSED** [review]

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/domain/entities/search/SortField.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package uk.gov.hmcts.reform.wataskmanagementapi.domain.entities.search;
+
+import com.fasterxml.jackson.annotation.JsonValue;
+
+public enum SortField {
+
+    DUE_DATE("dueDate"),
+    TASK_TITLE("taskTitle"),
+    LOCATION_NAME("locationName"),
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/domain/entities/search/SortField.java:9]: Looks like we no longer need to sort by **locationName**. Its crossed out on the ticket

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/domain/entities/camunda/CamundaSearchQuery.java [hunk 4]
```diff
@@ -58,6 +60,13 @@ public CamundaAndQueryBuilder andQuery(Map<String, List<String>> searchExpressio
             return this;
         }
 
+        public CamundaAndQueryBuilder andSortingQuery(List<CamundaSortingExpression> sortingExpressions) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/domain/entities/camunda/CamundaSearchQuery.java:63]: unit test to cover andSortingQuery method? CamundaSearchQueryTest?

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostTaskSearchControllerTest.java [hunk 5]
```diff
@@ -251,6 +258,156 @@ public void should_return_a_200_with_empty_search_results_user_jurisdiction_perm
...
+
+        tasksCreated.stream()
+            .map(map -> map.get("taskId"))
+            .forEach(taskId -> common.cleanUpTask(taskId, REASON_COMPLETED));
+    }
+
+
+    @Test
+    public void should_return_a_200_with_search_results_and_sorted_by_case_id_due_date() {
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostTaskSearchControllerTest.java:375]: should test name be renamed to qualify what test is doing? i.e search results and sorted by dueDate descendant

**Review by BenNzewi — APPROVED** [review]

**Review by lewis150193 — APPROVED** [review]

**Review by indvamshi — APPROVED** [review]
looks good.

**Pull request merged**